### PR TITLE
fix(query-bar, editor): update query history autocompletion to be more selective COMPASS-8241

### PR DESCRIPTION
**Pull request by Anemy**
COMPASS-8241

This pr restricts the query history auto completion items we show when a user is typing in the query bar.

Previously we would fuzzy match any characters in the entire query. That resulted in too many things being shown even when they weren't relevant. Now it'll only do this fuzzy matching when it the query history item passes a preliminary check that looks for if what the user is typing is a field prefix or contains a matching field + value in the query property.

Another part of this is not matching on the labels we provide for various fields. We used to match on the field property titles like 'project' and 'sort', which would produce false positives and more noise. To do this we use the `displayLabel` property of a completion item. Unfortunately this results in us losing the bolding on the characters that are matched. It should be something we can implement support for using the `getMatch`: https://codemirror.net/docs/ref/#autocomplete.Completion.displayLabel
I didn't include in this pr as we want this fix out sooner, I'll create ticket for that and link here.

### CODE REVIEW

File: packages/compass-query-bar/src/components/query-history/index.spec.tsx [hunk 1]
```diff
@@ -92,12 +92,12 @@ const renderQueryHistory = (basepath: string) => {
 
 describe('query-history', function () {
   let tmpDir: string;
-  before(function () {
-    tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'compass-query-history'));
+  before(async function () {
+    tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'compass-query-history'));
```
Comment (Anemy) [inline packages/compass-query-bar/src/components/query-history/index.spec.tsx:96]: drive-by for a warning fix.
```
(node:65374) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead\
```

File: packages/compass-editor/src/codemirror/query-history-autocompleter.ts [hunk 2]
```diff
@@ -21,11 +21,89 @@ export type SavedQuery = {
...
+        simplifyQueryStringForAutocomplete(fieldString);
+
+      if (fieldStringSimplified.startsWith(input)) {
+        return true;
+      }
+      const inputIndex = inputToMatch.indexOf(fieldStringSimplified);
+      if (inputIndex !== -1) {
+        inputToMatch = inputToMatch.replace(fieldStringSimplified, '');
+      }
+    }
```
Comment (Anemy) [inline packages/compass-editor/src/codemirror/query-history-autocompleter.ts:86]: I'm thinking to make the UX a touch better we would add another loop after this that checks for if what's remaining in the `inputToMatch` is a prefix of one of the query history items. Right now we're on the side of not showing it all the time which I think is okay. I could add a TODO here and create a ticket.

**Review by alenakhineika — APPROVED** [review]

File: packages/compass-editor/src/codemirror/query-history-autocompleter.ts [hunk 3]
```diff
@@ -21,11 +21,89 @@ export type SavedQuery = {
...
+      const queryValueSimplified =
+        simplifyQueryStringForAutocomplete(queryValueString);
+      return queryValueSimplified.startsWith(input);
+    }
+
+    // We subtract parts of the string until there's nothing left.
+    // We know it's a possible match if the all of the input is found in the query.
+    let inputToMatch = input;
+
+    // We check each top level field as they can be in any order.
```
Comment (gribnoysup) [inline packages/compass-editor/src/codemirror/query-history-autocompleter.ts:70]: If we're introducing a performance bottleneck for rendering with this logic, we should consider making autocompleter async and moving this code out of the main thread to prevent blocking operations on main that will have a negative effect on user interactions with the editor. Let's open a ticket for that if that's the case

File: packages/compass-editor/src/codemirror/query-history-autocompleter.ts [hunk 4]
```diff
@@ -21,11 +21,89 @@ export type SavedQuery = {
...
+    for (const [key, value] of Object.entries(queryValue).slice(
+      0,
+      30 /* Some queries can have a ton of fields, we slice to avoid long loops on each character typed. */
```
Comment (gribnoysup) [inline packages/compass-editor/src/codemirror/query-history-autocompleter.ts:73]: If we're introducing a performance bottleneck for rendering with this logic, we should consider making autocompleter async and moving this code out of the main thread to prevent blocking operations on main that will have a negative effect on user interactions with the editor. Let's open a ticket for that if that's the case

Comment (lerouxb) [reply packages/compass-editor/src/codemirror/query-history-autocompleter.ts:73]: Just keep in mind that async could mean that we'd have to be careful that we're not changing the results underneath the user. Like they could have started pressing enter in the meantime and then the results change again.

Comment (Anemy) [reply packages/compass-editor/src/codemirror/query-history-autocompleter.ts:73]: @gribnoysup The occurrences of 30 fields on a query are pretty rare, I also think it's even rarer that someone would then be typing a query of that scale and then want it to auto complete. I'd like to avoid too much early optimization here, although what I'm adding here may be doing that. I'm happy to open a ticket to add async autocompletion, however I think it wouldn't be impacting many folks here. Worst case the user doesn't see a certain query autocompleted with this UI.
Do you feel async autocompletion is something we'll want to do eventually? I'm mostly wondering if we can avoid creating a ticket we might not want to ever spend the time to implement and maintain so I'd like to hear your thought.

Comment (gribnoysup) [reply packages/compass-editor/src/codemirror/query-history-autocompleter.ts:73]: I'm going only based on what the comments in this code are saying: if this code is slow and because it's running every time user types any symbol in the input box, the worst case scenario is not missing items in the autocomplete, but getting a noticeable lag when typing in the input field. So if cutting the amount of fields here makes it fast enough so that it's not noticeable in terms of rendering performance, that's good enough 👍 

> Do you feel async autocompletion is something we'll want to do eventually?

I think we'll have to based on the upcoming scopes for autocompletion, but this is just me speculating at this point and should probably not affect the decision for this change 🙂

Comment (Anemy) [reply packages/compass-editor/src/codemirror/query-history-autocompleter.ts:73]: Sounds good, I appreciate the callout there, stoked for the autocompletion project.
